include/freerdp/utils/pod_arrays.h

```diff
@@ -0,0 +1,146 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * POD arrays
+ *
+ * Copyright 2022 David Fort <contact@hardening-consulting.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FREERDP_UTILS_POD_ARRAYS_H_
+#define FREERDP_UTILS_POD_ARRAYS_H_
+
+#include <winpr/wtypes.h>
+#include <freerdp/freerdp.h>
+
+#define POD_ARRAYS_DECL(T, TLOWER) \
+	typedef struct \
+	{ \
+		T* values; \
+		size_t nvalues; \
+	} Array##T; \
+	typedef BOOL Array##T##Cb(T v, void* data); \
+	\
+	FREERDP_API void array_##TLOWER##_init(Array##T* a); \
+	FREERDP_API size_t array_##TLOWER##_size(const Array##T* a); \
+	FREERDP_API T* array_##TLOWER##_data(const Array##T* a); \
+	FREERDP_API const T* array_##TLOWER##_cdata(const Array##T* a); \
+	FREERDP_API T array_##TLOWER##_get(const Array##T* a, size_t idx); \
+	FREERDP_API void array_##TLOWER##_set(Array##T* a, size_t idx, T v); \
+	FREERDP_API BOOL array_##TLOWER##_append(Array##T* a, T v); \
+	FREERDP_API BOOL array_##TLOWER##_contains(const Array##T* a, T v); \
+	FREERDP_API BOOL array_##TLOWER##_foreach(const Array##T* a, Array##T##Cb cb, void* data); \
+	FREERDP_API void array_##TLOWER##_reset(Array##T* a); \
+	FREERDP_API void array_##TLOWER##_uninit(Array##T* a)
+
+#define POD_ARRAYS_IMPL(T, TLOWER) \
```

Comment:
Better move to `.c` file as the implementation should not be public.

Reply:
Setting it public allows to define your own arrays of basic objects (like a struct).

Reply:
hmm, then it might be better to create this as header only type?
e.g. make all functions `static INLINE`

Reply:
Could do that, but it will multiply the implementations in all C files that are using them.

Reply:
well, as far as I understand it this is just a convenience wrapper for dynamic arrays, so it should be sufficient to have a generic (typed) template we can use, the implementation itself will not really change much.

You still get the type safety and unified (internal) consistency checks over separate implementations and `one place to change them all`